--- /home/ke/opt/trale.orig/startup.pl	2009-07-07 10:20:11.000000000 +0200
+++ /home/ke/opt/trale/startup.pl	2009-07-07 20:34:33.000000000 +0200
@@ -141,18 +141,35 @@
 % saving them to a file, or doing a diff on results
 % Note: This needs to be loaded before any other
 % pretty printing hooks in order to have any effect.
-:- ensure_loaded(trale_home(save_results)).    
+:- ensure_loaded(trale_home(save_results)).
+
+ensure_ghooks_loaded :-
+    % load pretty printing hooks for grisu interface:
+    ensure_loaded(trale_home(ghooks)),
+    % diff of two feature structures
+    ensure_loaded(trale_home(diff)),
+    % debug unifies two feature structures
+    ensure_loaded(trale_home(debug_unify)).
 
 % load code for the graphical interface, if needed
 :- (   environ('INTERFACE',text) ->  true
-   ;   % load pretty printing hooks for grisu interface:
-       ensure_loaded(trale_home(ghooks)),
-       % diff of two feature structures
-       ensure_loaded(trale_home(diff)),
-       % debug unifies two feature structures
-       ensure_loaded(trale_home(debug_unify))   
+   ;   ensure_ghooks_loaded
    ).
 
+:- dynamic tralesld_active/0.
+
+% load code for the graphical debugger, if needed
+:- environ('DEBUGINTERFACE','debuggraphics(true)')
+   -> (ensure_loaded(trale_home('debugger/tralesld_hooks')),
+       assert(tralesld_active),
+       ensure_ghooks_loaded)
+    ; true.
+
+:- environ('DEBUGINTERFACE','debuggraphics(false)')
+   -> (ensure_loaded(trale_home('debugger/tralesld_hooks')),
+       ensure_ghooks_loaded)
+    ; true.
+
 % tsdb support: not (yet) generally loaded, call load_tsdb/0 to load
 load_tsdb :- ensure_loaded(trale_home('tsdb/itsdb')),
              ensure_loaded(trale_home('tsdb/trale-itsdb')).
@@ -164,4 +181,3 @@
 
 % call the startup code
 :- trale_startup.
-
